### PR TITLE
faiss: init at 1.7.2

### PR DESCRIPTION
**Pull request by SomeoneSerge**
###### Description of changes

Adds `faiss`, a library for fast nearest-neighbour search.
The derivation can be built with cuda support and with python bindings.
The `checkPhase` does nothing, but `passthru.tests` include the pytest suite and a demo from the upstream.
For the python part, the derivation expects `pythonPackages` as a package set, rather than individual python dependencies.
Since the main derivation is outside the python package set, I also chose not to populate `propagatedBuildInputs` (otherwise they should contain `numpy`)

New attributes:

```
faiss
python3Packages.faiss
```

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/science/math/faiss/default.nix [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    } // lib.optionalAttrs pythonSupport {
+      pytest = pythonPackages.callPackage ./pytest-suite.nix { };
+    };
+  };
+
+  meta = with lib; {
+    description = "A library for efficient similarity search and clustering of dense vectors by Facebook Research";
+    homepage = "https://github.com/facebookresearch/faiss";
+    license = licenses.mit;
+    platforms = platforms.unix; # Never tested windows
```
Comment (samuela) [inline pkgs/development/libraries/science/math/faiss/default.nix:120]: I don't think nix supports windows anyhow

File: pkgs/development/libraries/science/math/faiss/pytest-suite.nix [hunk 2]
```diff
@@ -0,0 +1,30 @@
+{ lib
```
Comment (samuela) [inline pkgs/development/libraries/science/math/faiss/pytest-suite.nix:1]: Why is this file necessary? Can't the tests be handled in the main derivation?

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -33023,6 +33023,11 @@ with pkgs;
     jre = openjdk11;
   };
 
+  faiss = callPackage ../development/libraries/science/math/faiss {
+    pythonPackages = python3Packages;
+    swig = swig4;
```
Comment (samuela) [inline pkgs/top-level/all-packages.nix:33028]: Is there anything special about swig4? Why not just let it use the default swig?

Comment (SomeoneSerge) [reply pkgs/top-level/all-packages.nix:33028]: Yes, swig3 doesn't support `-doxygen` option that upstream hardcodes

File: pkgs/development/libraries/science/math/faiss/pytest-suite.nix (hunk 2 above)
Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/faiss/pytest-suite.nix:1]: I could write a custom `checkPhase ` instead, but since the linkage takes rather long for this package, I tried to untangle the derivation and the tests. I grepped for modules that potentially could do something similar first and there were e.g. matches for running `pytest` in `passthru.tests`.

On a side note, I think separating long tests from long builds is desirable for big packages, including some that are already in nixpkgs. I cannot defend the particular structure though

File: pkgs/development/libraries/science/math/faiss/default.nix (hunk 1 above)
Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/faiss/default.nix:120]: Faiss does though, and there's mingw, I guess?

File: pkgs/development/libraries/science/math/faiss/pytest-suite.nix (hunk 2 above)
Comment (samuela) [reply pkgs/development/libraries/science/math/faiss/pytest-suite.nix:1]: ok that makes sense to me... just was confusing at first glance!

File: pkgs/top-level/all-packages.nix (hunk 3 above)
Comment (samuela) [reply pkgs/top-level/all-packages.nix:33028]: gotcha, could we add a comment to that effect here?

File: pkgs/development/libraries/science/math/faiss/pytest-suite.nix (hunk 2 above)
Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/faiss/pytest-suite.nix:1]: I should be happy to rewrite this so as to reduce confusion! If only you got an idea

File: pkgs/development/libraries/science/math/faiss/default.nix (hunk 1 above)
Comment (onny) [reply pkgs/development/libraries/science/math/faiss/default.nix:120]: I guess what @samuela means is that Windows isn't in the platforms scope in NixOS. The platform part is only relevant for building on darwin and linux including the different linux architectures. So it is safe to remove the comment.

File: pkgs/development/libraries/science/math/faiss/pytest-suite.nix (hunk 2 above)
Comment (onny) [reply pkgs/development/libraries/science/math/faiss/pytest-suite.nix:1]: I'm also used this approach and put tests into a seperate file `tests.nix`to avoid infinite recursion, you can have a look here  https://github.com/NixOS/nixpkgs/tree/80ec2d1ace7c133ce640a3ebc72130b6c211039f/pkgs/development/python-modules/doit

**Comment by onny** [comment]
An other PR depends on this package. The work you did looks really great, hope it can get merged.

File: pkgs/development/libraries/science/math/faiss/default.nix (hunk 1 above)
Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/faiss/default.nix:120]: Sorry for a late response. I'll strip the comments and sanity-check the expression again before tomorrow

**Comment by SomeoneSerge** [comment]
No major changes in the last force-push, just:

- rebased on master,
- renamed tests.

I checked `nix-build -A python3Packages.faiss.tests` and `nix build .#faiss`

This may be squashed-n-merged now

**Comment by onny** [comment]
Unfortunately I'm unable to build and test it locally because the build process is eating my memory ^^ Not sure if this is normal but my system seems to weak for building it with nixpkg-review.

**Comment by SomeoneSerge** [comment]
Result of `nixpkgs-review pr 168197` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>faiss (python310Packages.faiss)</li>
    <li>python39Packages.faiss</li>
  </ul>
</details>

**Review by samuela — APPROVED** [review]
LGTM any hold up to merging?
